src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -1460,17 +1460,37 @@ GenTree* Lowering::LowerHWIntrinsicCmpOp(GenTreeHWIntrinsic* node, genTreeOps cm
     // Special case: "vec ==/!= zero_vector"
     if (!varTypeIsFloating(simdBaseType) && (op != nullptr) && (simdSize != 12))
     {
-        GenTree* cmp = op;
+        uint64_t scalarAndMask = UINT64_MAX;
+        GenTree* cmp           = op;
         if (simdSize != 8) // we don't need compression for Vector64
         {
+            CorInfoType pairwiseMaxType = CORINFO_TYPE_UINT;
+
+            // If op is "vec & cnsVec" where both u64 components in that cnsVec are the same (for both SIMD12 and
+            // SIMD16) then we'd better do this AND on top of TYP_LONG NI_AdvSimd_Extract in the end - it produces a
+            // more optimal codegen.
+            if (op->OperIsHWIntrinsic(NI_AdvSimd_And) && op->AsHWIntrinsic()->Op(2)->OperIs(GT_CNS_VEC))
```

Comment:
This doesn't have to be a constant right?

Just any `(x & y) == zero` or `(x & y) != zero` can be optimzied down to a `tst` (on both xarch and arm64).